tecken/download/tasks.py

```diff
@@ -0,0 +1,192 @@
+    EndpointConnectionError,
+    ConnectionError,
+    ClientError,
+))
+@reraise_clienterrors
+@reraise_endpointconnectionerrors
+def download_microsoft_symbol(symbol, debugid):
+    MS_URL = 'https://msdl.microsoft.com/download/symbols/'
+    MS_USER_AGENT = 'Microsoft-Symbol-Server/6.3.0.0'
+    debug_file = symbol
+    url = MS_URL + "/".join([debug_file, debugid, debug_file[:-1] + '_'])
+    session = requests_retry_session()
+    r = session.get(url, headers={'User-Agent': MS_USER_AGENT})
+    if r.status_code != 200:
+        logger.debug(
+            f'Symbol {debug_file}/{debugid} does not '
+            'exist on msdl.microsoft.com'
+        )
+        return
+
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        filepath = os.path.join(tmpdirname, os.path.basename(url))
+        with open(filepath, 'wb') as f:
+            content = r.content
+            if not content.startswith(b'MSCF'):
+                logger.info(
+                    f"Beginning of content in {url} did not start with 'MSCF'"
+                )
+                return
+            f.write(content)
+
+        cmd = [
+            settings.CABEXTRACT_PATH,
+            '--quiet',
+            # Important so that the extract .pdb filename is predictable.
+            '--lowercase',
+            '--directory', tmpdirname,
+            filepath,
+        ]
+        logger.debug(' '.join(cmd))
+        pipe = subprocess.Popen(
+            cmd,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE
+        )
+        with metrics.timer('cabextract'):
+            std_out, std_err = pipe.communicate()
+            if std_err:
+                logger.warning(
+                    f'cabextract failed for {url}. Error: {std_err!r}'
+                )
+                return
+
+        # Running cabextract creates a file 'foo.pdb' from 'foo.pd_'
+        pdb_filepath = filepath.lower().replace('.pd_', '.pdb')
+        assert pdb_filepath != filepath
+        assert os.path.isfile(pdb_filepath), pdb_filepath
+        cmd = [
+            settings.DUMP_SYMS_PATH,
+            pdb_filepath,
+        ]
+        logger.debug(' '.join(cmd))
+        pipe = subprocess.Popen(
+            cmd,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE
+        )
+        with metrics.timer('dump_syms'):
+            std_out, std_err = pipe.communicate()
+            # Note! It's expected, even if the dump_syms call works,
+            # that the stderr contains something like:
+            # b'Failed to find paired exe/dll file\n'
+            # which is fine and can be ignored.
+            if std_err and not std_out:
+                raise DumpSymsError(
+                    f'dump_syms extraction failed for {url}. '
+                    f'Error: {std_err!r}'
+                )
+
+        # Let's go ahead and upload it now, if it hasn't been uploaded
+        # before.
+        upload_microsoft_symbol(symbol, debugid, std_out)
+
```

Comment:
All this meat above is basically a fancy refactoring of https://gist.github.com/luser/92d5bc88478665554898

One important feature difference, that your gist didn't get right, is that it forces the generated `.pdb` file to be in lowercase. If you do `cabextract -d /tmp FOO.pd_` it will create `/tmp/foo.pdb`. Not `/tmp/FOO.pdb` so you can't just assume the same name by replacing the `_` suffix with a `d`.